### ISSUE TITLE
Magic `__get` methods interfere with classes that share the same parent

### ISSUE THREAD

**Issue by andrewandante** [silverstripe/silverstripe-framework#10263]
## Affected Version

4.10 (is what I've tested on)

## Description

I have in my project a `Page` and `PageController` (thanks, CMS)
I have two Page types:
- `CoolPage extends Page`
- `DopePage extends Page`

`CoolPage` has a `$db` field `Summary`
`DopePage` has a method `getSummary()`

Adding `$Summary` in my `CoolPage.ss` throws the error:

> [Emergency] Uncaught BadMethodCallException: Object->__call(): the method 'getSummary' does not exist on CoolPage

I've tracked this down to the `__get()` method on `ViewableData`, because the following line returns `true`:

```php
        if (strtolower($property) !== 'field' && $this->hasMethod($method = "get$property")) {
```

Which is because `PageController::hasMethod('getSummary')` returns `true`, as it's getting it from `DopePage`, but then it can't call it in the next line:

```php
        return $this->$method();
```

because `CoolPage`, the current scope, doesn't.

I've been able to work around this by creating an otherwise empty `CoolPageController extends PageController`, but that feels inelegant.

**Comment by kinglozzer** [comment]
> Which is because PageController::hasMethod('getSummary') returns true, as it's getting it from DopePage

Is `DopePage` added into the mix via `->customise()` or something? I’m not sure why `PageController` would ever be looking at `DopePage` when rendering a `CoolPage`

**Comment by andrewandante** [comment]
That I'm not totally sure about to be honest, but it seems to come from:

```php
    public function hasMethod($method)
    {
        return method_exists($this, $method) || $this->getExtraMethodConfig($method);
    }
```

calling (from `Extensible`):

```php
    protected function defineMethods()
    {
        $this->defineMethodsCustom();

        // Define extension methods
        $this->defineExtensionMethods();
    }
```

And it comes in through `defineExtensionMethods()` - maybe this has to do with the magic connection between `Page` and `PageController`? In which case it's a CMS issue

**Comment by NightJar** [comment]
This is really confusing to follow. Do you have some actual minimal reproduction code you could share instead? So far I read into it what Loz has stated already.

**Comment by andrewandante** [comment]
Sure thing.

```php
AndantePage.php

<?php

class AndantePage extends Page
{

    private static $db = [
        'Summary' => 'Varchar(255)',
    ];

}
```

```php
NightjarPage.php

<?php

class NightjarPage extends Page
{

    public function getSummary()
    {
        return 'Nightjar in the day is just a Dayjar';
    }

}
```

```html
AndantePage.ss

$Summary
```

Expected result when visiting any old `AndantePage`: The `Summary` field is rendered in the template.
Actual result: `[Emergency] Uncaught BadMethodCallException: Object->__call(): the method 'getSummary' does not exist on AndantePage`

**Comment by kinglozzer** [comment]
I can’t recreate this - I just get nothing rendered for `$Summary`, or if I add something to that field in `AndantePage` in the CMS it renders as expected. `NightjarPage` then renders `Nightjar in the day is just a Dayjar` as expected too

**Comment by christopherdarling** [comment]
Can't recreate this either on the latest release.
@andrewandante worth double checking you haven't overloaded `__get()` on page.php and/or have an extension applied to either page-type that's causing this?

**Comment by andrewandante** [comment]
How curious. I'll dig in a little and see if it's coming from somewhere else, though I'm pretty sure we aren't doing anything particularly weird in this project. Will see what exactly I need to set up to reproduce.

**Comment by andrewandante** [comment]
Ah-ha. The `AndantePage` has to be the child of the `NightjarPage`

**Comment by michalkleiner** [comment]
If by the child you mean nested underneath, that could suggest it's related to the `Hierarchy` class or thereabouts.

**Comment by andrewandante** [comment]
> nested underneath
Yes that's right. I suspect something to do with Scope and Hierarchy and the magic that lets Controllers pull methods from their Pages. More investigation required, but at least I can reproduce it and am not losing my mind entirely. Yet.

**Comment by kinglozzer** [comment]
Had a quick look into this, I think the problem comes from the `CustomMethods` trait. During request processing of a parent/child page structure, a `ContentController` is constructed with `NightjarPage`, and then a second one with `AndantePage`. The problem is that the second one built “inherits” the custom methods defined by the first one, despite being an entirely fresh instance of the controller. This is because the list of custom methods is static, so shared across multiple instances:

https://github.com/silverstripe/silverstripe-framework/blob/78c9aea61fc07f275cb84512a0d2f155613e13ac/src/Core/CustomMethods.php#L15-L20

After the controller is built for the parent page, it looks (roughly) like this:

```php
$extra_methods = [
    ContentController::class => [
        'getSummary'
    ]
];
```

The second instance of `ContentController` comes along and inherits this, because it’s also an instance of `ContentController`. `ViewableData` does actually try to clear these methods when `->setFailover($page)` is called, but it fails because it’s removing the child `AndantePage`’s methods instead of `NightjarPage`.

I’m not really sure what the best solution here is. Sharing these methods statically probably isn’t a good idea, but I’d be wary about performance regressions from changing that

**Comment by GuySartorelli** [comment]
Oh hey, that looks like https://github.com/silverstripe/silverstripe-framework/issues/11574
I'm gonna close this in favour of that one which strikes at the root cause.
